internal/broadcaster/utxo_creator.go

```diff
@@ -7,153 +7,149 @@ import (
+func (b *UTXOCreator) Shutdown() {
+	b.cancelAll()
+	b.wg.Wait()
+}
+
+func NewUTXOCreator(logger *slog.Logger, client ArcClient, keySet *keyset.KeySet, utxoClient UtxoClient, isTestnet bool, opts ...func(p *Broadcaster)) (*UTXOCreator, error) {
 	b, err := NewBroadcaster(logger, client, utxoClient, isTestnet, opts...)
 	if err != nil {
 		return nil, err
 	}
 
 	creator := &UTXOCreator{
 		Broadcaster: b,
-		keySets:     keySets,
+		keySet:      keySet,
 	}
 
 	return creator, nil
 }
 
 func (b *UTXOCreator) CreateUtxos(requestedOutputs int, requestedSatoshisPerOutput uint64) error {
-	for _, ks := range b.keySets {
-		b.logger.Info("creating utxos", slog.String("address", ks.Address(!b.isTestnet)))
+	b.logger.Info("creating utxos", slog.String("address", b.keySet.Address(!b.isTestnet)))
 
-		requestedOutputsSatoshis := int64(requestedOutputs) * int64(requestedSatoshisPerOutput)
+	requestedOutputsSatoshis := int64(requestedOutputs) * int64(requestedSatoshisPerOutput)
 
-		confirmed, unconfirmed, err := b.utxoClient.GetBalanceWithRetries(b.ctx, ks.Address(!b.isTestnet), 1*time.Second, 5)
-		if err != nil {
-			return err
-		}
+	confirmed, unconfirmed, err := b.utxoClient.GetBalanceWithRetries(b.ctx, b.keySet.Address(!b.isTestnet), 1*time.Second, 5)
+	if err != nil {
+		return err
+	}
 
-		balance := confirmed + unconfirmed
+	balance := confirmed + unconfirmed
 
-		if requestedOutputsSatoshis > balance {
-			return fmt.Errorf("requested total of satoshis %d exceeds balance on funding keyset %d", requestedOutputsSatoshis, balance)
-		}
+	if requestedOutputsSatoshis > balance {
+		return fmt.Errorf("requested total of satoshis %d exceeds balance on funding keyset %d", requestedOutputsSatoshis, balance)
+	}
 
-		utxos, err := b.utxoClient.GetUTXOsWithRetries(b.ctx, ks.Script, ks.Address(!b.isTestnet), 1*time.Second, 5)
-		if err != nil {
-			return err
+	utxos, err := b.utxoClient.GetUTXOsWithRetries(b.ctx, b.keySet.Script, b.keySet.Address(!b.isTestnet), 1*time.Second, 5)
+	if err != nil {
+		return err
+	}
+
+	utxoSet := list.New()
+	for _, utxo := range utxos {
+		if utxo.Satoshis >= requestedSatoshisPerOutput {
+			utxoSet.PushBack(utxo)
 		}
+	}
 
-		utxoSet := list.New()
-		for _, utxo := range utxos {
-			// collect right sized utxos
-			if utxo.Satoshis >= requestedSatoshisPerOutput {
-				utxoSet.PushBack(utxo)
-				continue
-			}
+	if utxoSet.Len() >= requestedOutputs {
+		b.logger.Info("utxo set", slog.Int("ready", utxoSet.Len()), slog.Int("requested", requestedOutputs), slog.Uint64("satoshis", requestedSatoshisPerOutput))
+		return nil
+	}
+
+	satoshiMap := map[string][]splittingOutput{}
+	lastUtxoSetLen := 0
+
+	for {
+		if lastUtxoSetLen >= utxoSet.Len() {
+			b.logger.Error("utxo set length hasn't changed since last iteration")
+			break
 		}
+		lastUtxoSetLen = utxoSet.Len()
 
-		// if requested outputs satisfied, return
```

Comment:
We shouldn't remove such comments as they can facilitate the understanding of the code

Reply:
Not sure why you resolved the conversation here because the comment is still missing

Reply:
I thought i pushed the change, somehow it went missing pushed again